HomeWork_11/HW11.py

```diff
@@ -0,0 +1,74 @@
+from datetime import datetime
+from time import perf_counter
+
+
+
+def decorator(func):
```

Comment:
Why do you implement here two decorators? Let's try to implement one, but make it works correctly?

So, the key point of decoration is that it takes a function as an argument, has a wrapper function inside and return this wrapper.
```
def deco(func):
    def wrapper():
        pass
    return wrapper
```
We return wrapper here and don't execute it since this wrapper will "replace" a function that will be decorated. So, the wrapper will be called when  a decorated function will be called and code inside the wrapper will be executed.

According to this, we should understand that our wrapper will replace our function so we need to adopt it to work with all function. Functions can take some arguments, but we don't know what exact argument they are, so we can use *args and **kwargs to fit all possible options:
```
def deco(func):
    def wrapper(*args, **kwargs):
        pass
    return wrapper
```

Next, we should think how to execute a decorated function itself. Since we have it as a decorator parameter, we can use it and call it. When we call it we can pass arguments taken by wrapper, to execute our function correctly. And, to cover cases with decoration functions that return some value we need to return a value of the function's execution result:
```
def deco(func):
    def wrapper():
        return func(*args, **kwargs)
    return wrapper
```

That's it, decorator is completed. Now, we can add some logs before we return from our wrapper:
def deco(func):
    def wrapper():
        print(f"This is some log. Function name is {func}, the time is {datetime.now()}")
        return func(*args, **kwargs)
    return wrapper